libcudacxx/docs/extended_api/memory_resource.md

```diff
@@ -0,0 +1,220 @@
+```c++
+struct valid_resource {
+  void* allocate(std::size_t, std::size_t) { return nullptr; }
+  void deallocate(void*, std::size_t, std::size_t) {}
+  void* allocate_async(std::size_t, std::size_t, cuda::stream_ref) { return nullptr; }
+  void deallocate_async(void*, std::size_t, std::size_t, cuda::stream_ref) {}
+  bool operator==(const valid_resource&) const { return true; }
+  bool operator!=(const valid_resource&) const { return false; }
+};
+static_assert(cuda::mr::async_resource<valid_resource>, "");
+```
+
+A library can now easily decide whether it wants to use the async interface:
+```c++
+template<class MemoryResource>
+    requires cuda::mr::resource<MemoryResource>
+void* maybe_allocate_async(MemoryResource& resource, std::size_t size, std::size_t align, cuda::stream_ref stream) {
+    if constexpr(cuda::mr::async_resource<MemoryResource>) {
+        return resource.allocate_async(size, align, stream);
+    } else {
+        return resource.allocate(size, align);
+    }
+}
+```
+
+### Putting them together
+
+Obviously, we want to combine type checks for arbitrary properties with the `{async_}resource` concept. This is what the `{async_}resource_with` concept is for. A library can now define arbitrary properties:
+```c++
+struct required_alignment{};
+struct my_memory_resource {
+    void* allocate(std::size_t, std::size_t) { return nullptr; }
+    void deallocate(void*, std::size_t, std::size_t) {}
+    bool operator==(const my_memory_resource&) const { return true; }
+    bool operator!=(const my_memory_resource&) const { return false; }
+
+    friend constexpr std::size_t get_property(const my_memory_resource& resource, required_alignment) noexcept { return resource.required_alignment; }
+
+    std::size_t required_alignment;
+};
+
+template<class MemoryResource>
+    requires cuda::mr::resource<MemoryResource>
+void* maybe_allocate_async_check_alignment(MemoryResource& resource, std::size_t size, cuda::stream_ref stream) {
+    if constexpr(cuda::mr::async_resource_with<MemoryResource, desired_alignment>) {
+        return resource.allocate_async(size, get_property(resource, required_alignment), stream);
+    } else if constexpr (cuda::mr::async_resource<MemoryResource>) {
+        return resource.allocate_async(size, my_default_alignment, stream);
+    } else if constexpr (cuda::mr::resource_with<MemoryResource, desired_alignment>) {
+        return resource.allocate(size, get_property(resource, required_alignment));
+    } else {
+        return resource.allocate(size, my_default_alignment);
+    }
+}
+
+// Potentially more consice
+template<class MemoryResource>
+    requires cuda::mr::resource<MemoryResource>
+void* maybe_allocate_async_check_alignment2(MemoryResource& resource, std::size_t size, cuda::stream_ref stream) {
```

Comment:
Question: is `cuda::stream_ref` documented?

Reply:
Nope, totally forgot about that